java-spiffe-core/src/main/java/spiffe/workloadapi/WorkloadApiClient.java

```diff
@@ -0,0 +1,240 @@
+    private final SpiffeWorkloadAPIBlockingStub workloadApiBlockingStub;
+    private final ManagedChannel managedChannel;
+    private final List<Context.CancellableContext> cancellableContexts;
+
+    private WorkloadApiClient(SpiffeWorkloadAPIStub workloadApiAsyncStub, SpiffeWorkloadAPIBlockingStub workloadApiBlockingStub, ManagedChannel managedChannel) {
+        this.workloadApiAsyncStub = workloadApiAsyncStub;
+        this.workloadApiBlockingStub = workloadApiBlockingStub;
+        this.managedChannel = managedChannel;
+        this.cancellableContexts = Collections.synchronizedList(new ArrayList<>());
+    }
+
+    /**
+     * Creates a new Workload API client using the default socket endpoint address.
+     * @see Address#getDefaultAddress()
+     *
+     * @return a {@link WorkloadApiClient}
+     */
+    public static WorkloadApiClient newClient() throws SocketEndpointAddressException {
+        val options = ClientOptions.builder().build();
+        return newClient(options);
+    }
+
+    /**
+     * Creates a new Workload API client.
+     * <p>
+     * If the SPIFFE socket endpoint address is not provided in the options, it uses the default address.
+     *
+     * @param options {@link ClientOptions}
+     * @return a {@link WorkloadApiClient}
+     */
+    public static WorkloadApiClient newClient(@NonNull ClientOptions options) throws SocketEndpointAddressException {
+        String spiffeSocketPath;
+        if (StringUtils.isNotBlank(options.spiffeSocketPath)) {
+            spiffeSocketPath = options.spiffeSocketPath;
+        } else {
+            spiffeSocketPath = Address.getDefaultAddress();
+        }
+
+        val socketEndpointAddress = Address.parseAddress(spiffeSocketPath);
+        val managedChannel = GrpcManagedChannelFactory.newChannel(socketEndpointAddress);
+
+        val workloadAPIAsyncStub = SpiffeWorkloadAPIGrpc
+                .newStub(managedChannel)
+                .withInterceptors(new SecurityHeaderInterceptor());
+
+        val workloadAPIBlockingStub = SpiffeWorkloadAPIGrpc
+                .newBlockingStub(managedChannel)
+                .withInterceptors(new SecurityHeaderInterceptor());
+
+        return new WorkloadApiClient(workloadAPIAsyncStub, workloadAPIBlockingStub, managedChannel);
+    }
+
+    /**
+     * One-shot blocking fetch call to get an X.509 context.
+     *
+     * @throws X509ContextException if there is an error fetching or processing the X.509 context
+     */
+    public X509Context fetchX509Context() {
+        Context.CancellableContext cancellableContext;
+        cancellableContext = Context.current().withCancellation();
+        X509Context result;
+        try {
+            result = cancellableContext.call(this::processX509Context);
+        } catch (Exception e) {
+            throw new X509ContextException("Error fetching X509Context", e);
+        }
+        // close connection
+        cancellableContext.close();
```

Comment:
Should this be called inside a finally clause?

Reply:
good catch, will put it in a try-with-resources